Self closing tag causes NPE in cvc_complex_type_2_1CodeAction.doCodeAction

At the root of lsp4xml (with [money.xsd](https://github.com/angelozerr/lsp4xml/pull/337/files#diff-3e61076f7173cbb44d7d2240e14ddcd3) present), create a file with :
```xml
<money xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="org.eclipse.lsp4xml/src/test/resources/xsd/money.xsd" currency="euros"> </money>
```
When self closing the money tag, an NPE is thrown:

```
Error - 3:22:55 p.m.] Apr 03, 2019 03:22:55 org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError()
Message: Internal error: java.lang.NullPointerException
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.NullPointerException
	at org.eclipse.lsp4xml.extensions.contentmodel.participants.codeactions.cvc_complex_type_2_1CodeAction.doCodeAction(cvc_complex_type_2_1CodeAction.java:42)
	at org.eclipse.lsp4xml.extensions.contentmodel.participants.ContentModelCodeActionParticipant.doCodeAction(ContentModelCodeActionParticipant.java:45)
	at org.eclipse.lsp4xml.services.XMLCodeActions.doCodeActions(XMLCodeActions.java:43)
	at org.eclipse.lsp4xml.services.XMLLanguageService.doCodeActions(XMLLanguageService.java:188)
	at org.eclipse.lsp4xml.XMLTextDocumentService.lambda$codeAction$15(XMLTextDocumentService.java:340)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	... 6 more

```

![Apr-03-2019 15-23-57](https://user-images.githubusercontent.com/148698/55506971-b8c15480-5624-11e9-9ec8-54ef17c63350.gif)

--- a/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMAttr.java
+++ b/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMAttr.java
@@ -234,6 +234,10 @@ public class DOMAttr extends DOMNode implements org.w3c.dom.Attr {
 		this.nodeAttrValue = nodeAttrValue;
 	}
 
+	public boolean valueContainsOffset(int offset) {
+		return nodeAttrValue != null && offset >= nodeAttrValue.getStart() && offset < nodeAttrValue.getEnd();
+	}
+
 	public boolean isIncluded(int offset) {
 		return DOMNode.isIncluded(getStart(), getEnd(), offset);
 	}

--- a/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMAttributeDeclaration.java
+++ b/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMAttributeDeclaration.java
@@ -31,6 +31,8 @@ public interface CMAttributeDeclaration {
 	
 	String getDocumentation();
 
+	String getValueDocumentation(String value);
+
 	/**
 	 * Returns true if the attribute is required and false otherwise.
 	 * 

--- a/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelHoverParticipant.java
+++ b/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelHoverParticipant.java
@@ -57,13 +57,48 @@ public class ContentModelHoverParticipant extends HoverParticipantAdapter {
 
 		try {
 			ContentModelManager contentModelManager = hoverRequest.getComponent(ContentModelManager.class);
-			
 			CMElementDeclaration cmElement = contentModelManager.findCMElement(attribute.getOwnerElement());
 			if (cmElement != null) {
 				String attributeName = attribute.getName();
 				CMAttributeDeclaration cmAttribute = cmElement.findCMAttribute(attributeName);
 				if (cmAttribute != null) {
 					String doc = cmAttribute.getDocumentation();
+					if (doc != null && doc.length() > 0) {
+						MarkupContent content = new MarkupContent();
+						content.setKind(MarkupKind.PLAINTEXT);
+						content.setValue(doc);
+						return new Hover(content);
+					}
+				}
+			}
+		} catch (CacheResourceDownloadingException e) {
+			return getCacheWarningHover(e);
+		}
+		return null;
+	}
+
+	@Override
+	public Hover onAttributeValue(IHoverRequest hoverRequest) throws Exception {
+		DOMAttr attribute = (DOMAttr) hoverRequest.getNode();
+
+		//Attempts to compute specifically for XSI related attributes since
+		//the XSD itself does not have enough information. Should create a mock XSD eventually.
+		Hover temp = XSISchemaModel.computeHoverResponse(attribute, hoverRequest);
+		if(temp != null) {
+			return temp;
+		}
+		
+		try {
+			ContentModelManager contentModelManager = hoverRequest.getComponent(ContentModelManager.class);
+			
+			CMElementDeclaration cmElement = contentModelManager.findCMElement(attribute.getOwnerElement());
+			if (cmElement != null) {
+				String attributeName = attribute.getName();
+				CMAttributeDeclaration cmAttribute = cmElement.findCMAttribute(attributeName);
+				
+				String attributeValue = attribute.getValue();
+				if (cmAttribute != null) {
+					String doc = cmAttribute.getValueDocumentation(attributeValue);
 					if (doc != null && doc.length() > 0) {
 						MarkupContent content = new MarkupContent();
 						content.setKind(MarkupKind.PLAINTEXT);

--- a/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/codeactions/cvc_complex_type_2_1CodeAction.java
+++ b/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/codeactions/cvc_complex_type_2_1CodeAction.java
@@ -39,7 +39,13 @@ public class cvc_complex_type_2_1CodeAction implements ICodeActionParticipant {
 			DOMNode node = document.findNodeAt(offset);
 			if (node != null && node.isElement()) {
 				DOMElement element = (DOMElement) node;
-				int startOffset = element.getStartTagCloseOffset();
+				int startOffset;
+				if(element.isSelfClosed()) {
+					startOffset = element.getEnd();
+				}
+				else {
+					startOffset = element.getStartTagCloseOffset();
+				}
 				int endOffset = element.getEnd();
 				Range diagnosticRange = XMLPositionUtility.createRange(startOffset, endOffset, document);
 				CodeAction removeContentAction = CodeActionFactory.replace("Set element as empty", diagnosticRange,

--- a/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/dtd/contentmodel/CMDTDAttributeDeclaration.java
+++ b/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/dtd/contentmodel/CMDTDAttributeDeclaration.java
@@ -53,4 +53,9 @@ public class CMDTDAttributeDeclaration extends XMLAttributeDecl implements CMAtt
 		return super.simpleType.defaultType == XMLSimpleType.DEFAULT_TYPE_REQUIRED;
 	}
 
+	@Override
+	public String getValueDocumentation(String value) {
+		return null;
+	}
+
 }

--- a/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java
+++ b/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java
@@ -15,12 +15,15 @@ import java.util.Collections;
 
 import org.apache.xerces.impl.dv.xs.XSSimpleTypeDecl;
 import org.apache.xerces.impl.xs.XSComplexTypeDecl;
+import org.apache.xerces.xs.StringList;
 import org.apache.xerces.xs.XSAttributeDeclaration;
 import org.apache.xerces.xs.XSAttributeUse;
+import org.apache.xerces.xs.XSMultiValueFacet;
 import org.apache.xerces.xs.XSObjectList;
 import org.apache.xerces.xs.XSSimpleTypeDefinition;
 import org.apache.xerces.xs.XSTypeDefinition;
 import org.apache.xerces.xs.XSValue;
+import org.apache.xerces.xs.datatypes.ObjectList;
 import org.eclipse.lsp4xml.extensions.contentmodel.model.CMAttributeDeclaration;
 
 /**
@@ -63,6 +66,17 @@ public class CMXSDAttributeDeclaration implements CMAttributeDeclaration {
 		return documentation;
 	}
 
+	@Override
+	public String getValueDocumentation(String value) {
+		if (documentation != null) {
+			return documentation;
+		}
+		// Try get xs:annotation from the element declaration or type
+		XSObjectList annotations = getValueAnnotations();
+		documentation = XSDAnnotationModel.getDocumentation(annotations, value);
+		return documentation;
+	}
+
 	/**
 	 * Returns list of xs:annotation from the element declaration or type
 	 * declaration.
@@ -74,6 +88,7 @@ public class CMXSDAttributeDeclaration implements CMAttributeDeclaration {
 		// Try get xs:annotation from the element declaration
 		XSAttributeDeclaration attributeDeclaration = getAttrDeclaration();
 		XSObjectList annotation = attributeDeclaration.getAnnotations();
+		
 		if (annotation != null && annotation.getLength() > 0) {
 			return annotation;
 		}
@@ -88,6 +103,69 @@ public class CMXSDAttributeDeclaration implements CMAttributeDeclaration {
 			return ((XSSimpleTypeDecl) typeDefinition).getAnnotations();
 		}
 		return null;
+	}
+
+	/**
+	 * Returns list of xs:annotation from the element declaration or type
+	 * declaration.
+	 * 
+	 * Indicated by:
+	 * https://msdn.microsoft.com/en-us/library/ms256143(v=vs.110).aspx
+	 * xs:attribute tags have content of either an xs:annotation or xs:simpleType
+	 * 
+	 * @return list of xs:annotation from the element declaration or type
+	 *         declaration.
+	 */
+	private XSObjectList getValueAnnotations() {
+		// Try get xs:annotation from the element declaration
+		XSAttributeDeclaration attributeDeclaration = getAttrDeclaration();
+		XSSimpleTypeDefinition simpleTypeDefinition = attributeDeclaration.getTypeDefinition();
+		XSSimpleTypeDecl simpleTypeDecl;
+		
+	
+		XSObjectList annotation = null; // The XSD tag that holds the documentation tag
+
+		if(simpleTypeDefinition instanceof XSSimpleTypeDecl) {
+			simpleTypeDecl = (XSSimpleTypeDecl) simpleTypeDefinition;
+			XSObjectList multiFacets = simpleTypeDecl.getMultiValueFacets();
+			if(!multiFacets.isEmpty()) {
+				XSMultiValueFacet facet = (XSMultiValueFacet) multiFacets.get(0);
+				multiFacets = facet.getAnnotations();
+				Object[] annotationArray = multiFacets.toArray();
+				if(!onlyContainsNull(annotationArray)) { // if multiValueFacets has annotations
+					annotation = simpleTypeDecl.getMultiValueFacets();
+				}
+			}
+		}
+		if(annotation == null){ // There was no specific documentation for the value, so use the general attribute documentation
+			annotation = attributeDeclaration.getAnnotations();
+		}
+		if (annotation != null && annotation.getLength() > 0) {
+			return annotation;
+		}
+		// Try get xs:annotation from the type of element declaration
+		XSTypeDefinition typeDefinition = attributeDeclaration.getTypeDefinition();
+		if (typeDefinition == null) {
+			return null;
+		}
+		if (typeDefinition.getTypeCategory() == XSTypeDefinition.COMPLEX_TYPE) {
+			return ((XSComplexTypeDecl) typeDefinition).getAnnotations();
+		} else if (typeDefinition.getTypeCategory() == XSTypeDefinition.SIMPLE_TYPE) {
+			return ((XSSimpleTypeDecl) typeDefinition).getAnnotations();
+		}
+		return null;
+	}
+
+	private boolean onlyContainsNull(Object[] arr) {
+		if(arr == null || arr.length == 0) {
+			return true;
+		}
+		for (Object o : arr) {
+			if(o != null) {
+				return false;
+			}
+		}
+		return true;
 	}
 
 	@Override

--- a/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/XSDAnnotationModel.java
+++ b/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/XSDAnnotationModel.java
@@ -17,8 +17,11 @@ import java.io.StringReader;
 import javax.xml.parsers.SAXParser;
 import javax.xml.parsers.SAXParserFactory;
 
+import org.apache.xerces.impl.dv.ValidatedInfo;
 import org.apache.xerces.xs.XSAnnotation;
+import org.apache.xerces.xs.XSMultiValueFacet;
 import org.apache.xerces.xs.XSObjectList;
+import org.apache.xerces.xs.datatypes.ObjectList;
 import org.xml.sax.Attributes;
 import org.xml.sax.InputSource;
 import org.xml.sax.SAXException;
@@ -47,12 +50,36 @@ class XSDAnnotationModel {
 	}
 
 	public static String getDocumentation(XSObjectList annotations) {
+		return getDocumentation(annotations, null);
+	}
+
+	public static String getDocumentation(XSObjectList annotations, String value) {
 		if (annotations == null) {
 			return "";
 		}
 		StringBuilder doc = new StringBuilder();
 		for (Object object : annotations) {
-			XSAnnotation annotation = (XSAnnotation) object;
+			XSAnnotation annotation = null;
+			if(object instanceof XSMultiValueFacet && value != null) {
+				XSMultiValueFacet multiValueFacet = (XSMultiValueFacet) object;
+				ObjectList enumerationValues = multiValueFacet.getEnumerationValues();
+				XSObjectList annotationValues = multiValueFacet.getAnnotations();
+				for (int i = 0; i < enumerationValues.getLength(); i++) {
+					Object enumValue = enumerationValues.get(i);
+
+					//Assuming always ValidatedInfo
+					String enumString = ((ValidatedInfo) enumValue).stringValue();
+				
+					if(value.equals(enumString)) {
+						annotation = (XSAnnotation) annotationValues.get(i);
+						break;
+					}
+				}
+			}
+			else if(object instanceof XSAnnotation) {
+				annotation = (XSAnnotation) object;
+			}
+			
 			XSDAnnotationModel annotationModel = XSDAnnotationModel.load(annotation);
 			if (annotationModel != null) {
 				if (annotationModel.getAppInfo() != null) {

--- a/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHover.java
+++ b/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHover.java
@@ -17,6 +17,7 @@ import org.eclipse.lsp4j.Hover;
 import org.eclipse.lsp4j.Position;
 import org.eclipse.lsp4j.Range;
 import org.eclipse.lsp4xml.commons.BadLocationException;
+import org.eclipse.lsp4xml.dom.DOMAttr;
 import org.eclipse.lsp4xml.dom.DOMDocument;
 import org.eclipse.lsp4xml.dom.DOMElement;
 import org.eclipse.lsp4xml.dom.DOMNode;
@@ -69,8 +70,12 @@ class XMLHover {
 				return getTagHover(hoverRequest, tagRange, true);
 			}
 		} else if (node.isAttribute()) {
+			DOMAttr attr = (DOMAttr) node;
+			if(attr.valueContainsOffset(offset)) {
+				return getAttrValueHover(hoverRequest, null);
+			}
 			// Attribute is hover
-			return getAttrHover(hoverRequest, null);
+			return getAttrNameHover(hoverRequest, null);
 		}
 		return null;
 	}
@@ -126,12 +131,35 @@ class XMLHover {
 	 * @param attrRange     the attribute  range
 	 * @return the LSP hover from the hovered attribute.
 	 */
-	private Hover getAttrHover(HoverRequest hoverRequest, Range attrRange) {
+	private Hover getAttrNameHover(HoverRequest hoverRequest, Range attrRange) {
 		//hoverRequest.setTagRange(tagRange);
 		//hoverRequest.setOpen(open);
 		for (IHoverParticipant participant : extensionsRegistry.getHoverParticipants()) {
 			try {
 				Hover hover = participant.onAttributeName(hoverRequest);
+				if (hover != null) {
+					return hover;
+				}
+			} catch (Exception e) {
+				LOGGER.log(Level.SEVERE, "While performing IHoverParticipant#onTag", e);
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Returns the LSP hover from the hovered attribute.
+	 * 
+	 * @param hoverRequest the hover request.
+	 * @param attrRange     the attribute  range
+	 * @return the LSP hover from the hovered attribute.
+	 */
+	private Hover getAttrValueHover(HoverRequest hoverRequest, Range attrRange) {
+		//hoverRequest.setTagRange(tagRange);
+		//hoverRequest.setOpen(open);
+		for (IHoverParticipant participant : extensionsRegistry.getHoverParticipants()) {
+			try {
+				Hover hover = participant.onAttributeValue(hoverRequest);
 				if (hover != null) {
 					return hover;
 				}

--- a/org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaDiagnosticsTest.java
+++ b/org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaDiagnosticsTest.java
@@ -302,6 +302,15 @@ public class XMLSchemaDiagnosticsTest {
 		testDiagnosticsFor(xml, d);
 		testCodeActionsFor(xml, d, ca(d, te(5, 25, 5, 38, "/>")));
 	}
+
+	@Test
+	public void cvc_complex_type_2_1_SelfClosing() throws Exception {
+		String xml = "<money xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:noNamespaceSchemaLocation=\"src/test/resources/xsd/money.xsd\" currency=\"euros\"> </money>";
+		
+		Diagnostic d = d(0, 143, 0, 144, XMLSchemaErrorCode.cvc_complex_type_2_1);
+		testDiagnosticsFor(xml, d);
+		testCodeActionsFor(xml, d, ca(d, te(0, 142, 0, 152, "/>")));
+	}
 	
 	@Test
 	public void cvc_complex_type_2_1WithLinefeed() throws Exception {

--- a/org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java
+++ b/org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java
@@ -30,7 +30,6 @@ public class XMLSchemaHoverExtensionsTest {
 		assertHover(xml,
 				"Defines a single (usually named) bean. A bean definition may contain nested tags for constructor arguments, property values, lookup methods, and replaced methods. Mixing constructor injection and setter injection on the same bean is explicitly supported.",
 				2);
-
 	};
 
 	@Test
@@ -146,6 +145,40 @@ public class XMLSchemaHoverExtensionsTest {
 		
 		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/invoice.xml",
 				null, null);
+	};
+
+	@Test
+	public void testHoverAttributeValueEuro() throws BadLocationException {
+		String xml = 
+		"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n" +
+		"<money xmlns=\"http://money\" currency=\"eu|ros\"\r\n" + // <- Hover
+		"	xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n" +
+		"	xsi:schemaLocation=\"http://money xsd/money.xsd\"></money>";
+		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/money.xml",
+				"Euro Hover", null);
+	};
+
+	@Test
+	public void testHoverAttributeValuePound() throws BadLocationException {
+		String xml = 
+		"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n" +
+		"<money xmlns=\"http://money\" currency=\"pou|nds\"\r\n" + // <- Hover
+		"	xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n" +
+		"	xsi:schemaLocation=\"http://money xsd/money.xsd\"></money>";
+		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/money.xml",
+				"Pound Hover", null);
+	};
+
+
+	@Test
+	public void testHoverAttributeValueNonExistent() throws BadLocationException {
+		String xml = 
+		"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n" +
+		"<money xmlns=\"http://money\" curr|ency=\"pounds\"\r\n" + // <- Hover
+		"	xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n" +
+		"	xsi:schemaLocation=\"http://money xsd/money.xsd\"></money>";
+		XMLAssert.assertHover(new XMLLanguageService(), xml, null, "src/test/resources/money.xml",
+				"Currency name Hover", null);
 	};
 
 	private static void assertHover(String value, String expectedHoverLabel, Integer expectedHoverOffset)

--- a/org.eclipse.lsp4xml/src/test/resources/catalogs/catalog.xml
+++ b/org.eclipse.lsp4xml/src/test/resources/catalogs/catalog.xml
@@ -23,6 +23,8 @@
 		
 	<system systemId="http://invoice.xsd" uri="../xsd/invoice.xsd" />
 
+	<system systemId="http://money.xsd" uri="../xsd/money.xsd" />
+
 	<uri name="http://docs.oasis-open.org/odata/ns/edmx"
 		uri="../xsd/edmx.xsd" />
 	<uri name="http://docs.oasis-open.org/odata/ns/edm"

--- a/org.eclipse.lsp4xml/src/test/resources/xsd/money.xsd
+++ b/org.eclipse.lsp4xml/src/test/resources/xsd/money.xsd
@@ -1,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified">
+
+<!-- definition of simpleTypes -->
+
+<xs:simpleType name="curr">
+  <xs:restriction base="xs:string">
+    <xs:enumeration value="pounds" >
+      <xs:annotation>
+        <xs:documentation>Pound Hover</xs:documentation>
+      </xs:annotation>
+    </xs:enumeration>
+    <xs:enumeration value="euros" >
+    <xs:annotation>
+        <xs:documentation>Euro Hover</xs:documentation>
+      </xs:annotation>
+    </xs:enumeration>
+    <xs:enumeration value="dollars" >
+      <xs:annotation>
+        <xs:documentation>Dollar Hover</xs:documentation>
+      </xs:annotation>
+    </xs:enumeration>
+  </xs:restriction>
+</xs:simpleType>
+
+<xs:element name="money" >
+        <xs:complexType>
+              <xs:attribute name="currency" type="curr" use="required" >
+                <xs:annotation>
+                  <xs:documentation>Currency name Hover</xs:documentation>
+                </xs:annotation>
+              </xs:attribute>
+        </xs:complexType>
+</xs:element>
+
+</xs:schema>

